confluent_kafka/schema_registry/protobuf.py

```diff
@@ -0,0 +1,437 @@
+    | record_subject_name_strategy         | {record name}                |
+    +--------------------------------------+------------------------------+
+
+    See ``Subject name strategy`` for additional details.
+
+    Args:
+        schema_registry_client (SchemaRegistryClient): Schema Registry
+            client instance.
+
+        descriptor (Descriptor): Protobuf Message Descriptor.
+
+        conf (dict): ProtobufSerializer configuration.
+
+    .. _Subject name strategy:
+        https://docs.confluent.io/current/schema-registry/serializer-formatter.html#subject-name-strategy
+
+    .. _Protobuf API reference:
+        https://googleapis.dev/python/protobuf/latest/google/protobuf.html
+
+    """  # noqa: E501
+    __slots__ = ['_auto_register', '_registry', '_known_subjects',
+                 'msg_index', '_proto_file', '_schema', '_schema_id',
+                 '_ref_reference_subject_func', '_subject_name_func']
+    # default configuration
+    _default_conf = {
+        'auto.register.schemas': True,
+        'subject.name.strategy': topic_subject_name_strategy,
+        'reference.subject.name.strategy': reference_subject_name_strategy
+    }
+
+    def __init__(self, schema_registry_client, descriptor, conf=None):
+        self._registry = schema_registry_client
+        self._schema_id = None
+        # Avoid calling registry if schema is known to be registered
+        self._known_subjects = set()
+
+        # handle configuration
+        conf_copy = self._default_conf.copy()
+        if conf is not None:
+            conf_copy.update(conf)
+
+        self._auto_register = conf_copy.pop('auto.register.schemas')
+        if not isinstance(self._auto_register, bool):
+            raise ValueError("auto.register.schemas must be a boolean value")
+
+        self._subject_name_func = conf_copy.pop('subject.name.strategy')
+        if not callable(self._subject_name_func):
+            raise ValueError("subject.name.strategy must be callable")
+
+        self._ref_reference_subject_func = conf_copy.pop('reference.subject.name.strategy')
+        if not callable(self._ref_reference_subject_func):
+            raise ValueError("subject.name.strategy must be callable")
+
+        if len(conf_copy) > 0:
+            raise ValueError("Unrecognized properties: {}"
+                             .format(conf_copy.keys()))
+
+        proto_file = descriptor.file
+
+        self._proto_file = proto_file
+        self.msg_index = _create_msg_index(descriptor)
+        self._schema = Schema(_schema_to_str(proto_file),
+                              schema_type='PROTOBUF')
+
+    @staticmethod
+    def _encode_index(buf, index):
+        """
+        Writes the index to buf formatted as unsigned varints.
+
+        Args:
+            buf (BytesIO): buffer to write index to
+            index [index]: Message index
+
+        Returns:
+            [bytes]: Serialized Message index
+
+        """
+        for value in index:
```

Comment:
I might be tired, but this varint encoder does not seem correct, it writes the continuation bit for the last octet if it is >0. shouldn't it be `while (value & ~0x7f): ` ?

Reply:
you are right, will fix encoder/decoder and add more tests.